File: autoresearch-guard/hooks/hook_runtime.py
from __future__ import annotations

import json
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any

SCRIPT_DIR = Path(__file__).resolve().parent
PLUGIN_ROOT = SCRIPT_DIR.parent
COMMON = PLUGIN_ROOT / "skills" / "autoresearch-guard" / "scripts"
sys.path.insert(0, str(COMMON))


@dataclass(frozen=True)
class HookEvent:
    payload: dict[str, Any]

    @property
    def session_id(self) -> str:
        return str(self.payload.get("session_id") or "")

    @property
    def turn_id(self) -> str:
        return str(self.payload.get("turn_id") or self.payload.get("prompt_id") or "")

    @property
    def tool_use_id(self) -> str:
        return str(self.payload.get("tool_use_id") or "")

    @property
    def tool_name(self) -> str:
        return str(self.payload.get("tool_name") or "")

    @property
    def cwd(self) -> Path:
        return Path(str(self.payload.get("cwd") or Path.cwd())).resolve()

    @property
    def command(self) -> str:
        tool_input = self.payload.get("tool_input")
        if isinstance(tool_input, dict):
            command = tool_input.get("command")
            if command:
                return str(command)
        return str(self.payload.get("command") or self.payload.get("cmd") or "")

    @property
    def stop_hook_active(self) -> bool:
        return self.payload.get("stop_hook_active") is True


def _stdin_payload_text(*, command: str = "") -> str:
    """Read Codex JSON from stdin without hanging CLI/tests that leave stdin open."""
    if sys.stdin.isatty():
        return ""
    # Prefer a non-blocking peek so subprocess tests without input= do not hang.
    try:
        import select

        ready, _, _ = select.select([sys.stdin], [], [], 0)
        if not ready:
            return ""
    except (ImportError, OSError, ValueError):
        if command:
            return ""
    return sys.stdin.read().strip()


def read_event(*, cwd: str = "", command: str = "") -> HookEvent:
    payload: dict[str, Any] = {}
    text = _stdin_payload_text(command=command)
    if text:
        try:
            loaded = json.loads(text)
        except json.JSONDecodeError:
            loaded = {"command": text, "malformed_input": True}
        if isinstance(loaded, dict):
            payload.update(loaded)
    if cwd:
        payload["cwd"] = cwd
    if command:
        payload.setdefault("tool_name", "Bash")
        payload["tool_input"] = {"command": command}
    return HookEvent(payload)


# Stop upward search at these markers so a package without .research does not
# inherit an ancestor campaign's hooks.
PROJECT_BOUNDARY_MARKERS = (
    ".git",
    ".arx-boundary",
    "package.json",
    "pyproject.toml",
    "Cargo.toml",
    "go.mod",
    "pom.xml",
    "composer.json",
    "Gemfile",
    "mix.exs",
)


def is_project_boundary(path: Path) -> bool:
    return any((path / name).exists() for name in PROJECT_BOUNDARY_MARKERS)


def find_research_root(cwd: Path) -> Path | None:
    resolved = Path(cwd).resolve()
    for path in [resolved, *resolved.parents]:
        if (path / ".research" / "current").exists():
            return path / ".research"
        if is_project_boundary(path):
            return None
    return None


def additional_context(event_name: str, message: str) -> dict[str, Any]:
    return {
        "hookSpecificOutput": {
            "hookEventName": event_name,
            "additionalContext": message,
        }
    }


def system_warning(message: str, *, stop: bool = False) -> dict[str, Any]:
    payload: dict[str, Any] = {"systemMessage": message}
    if stop:
        payload.update({"continue": False, "stopReason": message})
    return payload


def emit(payload: dict[str, Any] | None) -> None:
    if payload:
        print(json.dumps(payload, ensure_ascii=True, sort_keys=True))

File: autoresearch-guard/hooks/hook_runtime.js
"use strict";

const fs = require("fs");
const path = require("path");
const { spawnSync } = require("child_process");

const PROJECT_BOUNDARY_MARKERS = [
  ".git",
  ".arx-boundary",
  "package.json",
  "pyproject.toml",
  "Cargo.toml",
  "go.mod",
  "pom.xml",
  "composer.json",
  "Gemfile",
  "mix.exs",
];

function readStdinSync() {
  if (process.stdin.isTTY) {
    return "";
  }
  try {
    return fs.readFileSync(0, "utf8").trim();
  } catch {
    return "";
  }
}

function readEvent({ cwd = "", command = "" } = {}) {
  let payload = {};
  const text = readStdinSync();
  if (text) {
    try {
      const loaded = JSON.parse(text);
      if (loaded && typeof loaded === "object" && !Array.isArray(loaded)) {
        payload = { ...loaded };
      }
    } catch {
      payload = { command: text, malformed_input: true };
    }
  }
  if (cwd) {
    payload.cwd = cwd;
  }
  if (command) {
    if (!payload.tool_name) {
      payload.tool_name = "Bash";
    }
    payload.tool_input = { command };
  }
  return {
    payload,
    get session_id() {
      return String(payload.session_id || "");
    },
    get turn_id() {
      return String(payload.turn_id || payload.prompt_id || "");
    },
    get cwd() {
      return path.resolve(String(payload.cwd || process.cwd()));
    },
    get stop_hook_active() {
      return payload.stop_hook_active === true;
    },
  };
}

function isProjectBoundary(dir) {
  return PROJECT_BOUNDARY_MARKERS.some((name) => fs.existsSync(path.join(dir, name)));
}

function findResearchRoot(cwd) {
  let current = path.resolve(cwd);
  const { root } = path.parse(current);
  while (true) {
    if (fs.existsSync(path.join(current, ".research", "current"))) {
      return path.join(current, ".research");
    }
    if (isProjectBoundary(current)) {
      return null;
    }
    if (current === root) {
      return null;
    }
    const parent = path.dirname(current);
    if (parent === current) {
      return null;
    }
    current = parent;
  }
}

function additionalContext(eventName, message) {
  return {
    hookSpecificOutput: {
      hookEventName: eventName,
      additionalContext: message,
    },
  };
}

function systemWarning(message, { stop = false } = {}) {
  const payload = { systemMessage: message };
  if (stop) {
    payload.continue = false;
    payload.stopReason = message;
  }
  return payload;
}

function dumpJson(value) {
  if (value === null) {
    return "null";
  }
  if (typeof value === "boolean" || typeof value === "number") {
    return String(value);
  }
  if (typeof value === "string") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map(dumpJson).join(", ")}]`;
  }
  if (typeof value === "object") {
    const keys = Object.keys(value).sort();
    const body = keys.map((key) => `${JSON.stringify(key)}: ${dumpJson(value[key])}`).join(", ");
    return `{${body}}`;
  }
  return JSON.stringify(value);
}

function emit(payload) {
  if (payload) {
    process.stdout.write(`${dumpJson(payload)}\n`);
  }
}

function resolvePython() {
  if (process.env.ARX_PYTHON) {
    return [process.env.ARX_PYTHON];
  }
  const candidates =
    process.platform === "win32"
      ? [
          ["py", "-3"],
          ["python3"],
          ["python"],
        ]
      : [
          ["python3"],
          ["python"],
        ];
  for (const cmd of candidates) {
    const result = spawnSync(cmd[0], [...cmd.slice(1), "-c", "import sys; print(sys.executable)"], {
      encoding: "utf8",
      timeout: 8000,
      windowsHide: true,
    });
    if (result.status === 0) {
      const exe = String(result.stdout || "")
        .trim()
        .split(/\r?\n/)[0];
      if (exe) {
        return [exe];
      }
    }
  }
  throw new Error(
    "Python 3 not found for AutoResearch Guard hooks (tried python3/python/py -3). Set ARX_PYTHON to a Python 3 executable."
  );
}

function runBridge(op, { researchRoot, event, allowIncomplete = false, cwdProvided = false } = {}) {
  const bridge = path.join(__dirname, "arx_bridge.py");
  const python = resolvePython();
  const args = [bridge, op, "--research-root", researchRoot];
  if (allowIncomplete) {
    args.push("--allow-incomplete");
  }
  if (cwdProvided) {
    args.push("--cwd-provided");
  }
  const result = spawnSync(python[0], args, {
    input: JSON.stringify(event || {}),
    encoding: "utf8",
    timeout: 15000,
    windowsHide: true,
    env: process.env,
  });
  if (result.error) {
    throw result.error;
  }
  if (result.status !== 0) {
    const err = String(result.stderr || result.stdout || "").trim();
    throw new Error(err || `arx_bridge.py ${op} failed with code ${result.status}`);
  }
  const text = String(result.stdout || "").trim();
  if (!text) {
    return { skip: true };
  }
  return JSON.parse(text);
}

module.exports = {
  readEvent,
  findResearchRoot,
  additionalContext,
  systemWarning,
  emit,
  runBridge,
  resolvePython,
};

if (require.main === module) {
  const idx = process.argv.indexOf("--find-root");
  if (idx !== -1) {
    const target = process.argv[idx + 1] || process.cwd();
    const root = findResearchRoot(target);
    process.stdout.write(root || "");
    process.exit(0);
  }
  process.stderr.write("usage: node hook_runtime.js --find-root <cwd>\n");
  process.exit(2);
}
